src/lib.rs

```diff
@@ -4709,21 +4709,51 @@ macro_rules! include_value {
 ///
 /// Implementations of `ByteSlice` must promise that their implementations of
 /// [`Deref`] and [`DerefMut`] are "stable". In particular, given `B: ByteSlice`
-/// and `b: B`, `b` must always dereference to a byte slice with the same
-/// address and length. This is true for both `b.deref()` and `b.deref_mut()`.
-/// If `B: Copy` or `B: Clone`, then the same is also true of copies or clones
-/// of `b`. For example, `b.deref_mut()` must return a byte slice with the same
-/// address and length as `b.clone().deref()`.
+/// and `b: B`, two calls, each to either `b.deref()` or `b.deref_mut()`, must
+/// return a byte slice with the same address and length. This must hold even if
+/// the two calls are separated by an arbitrary sequence of calls to methods on
+/// `ByteSlice`, [`ByteSliceMut`], [`IntoByteSlice`], or [`IntoByteSliceMut`].
+/// This does *not* need to hold if the two calls are separated by any method
+/// calls, field accesses, or field modifications *other than* those from these
+/// traits.
+///
+/// Note that this also implies that, given `b: B`, the address and length
+/// cannot be modified via objects other than `b`, either on the same thread or
+/// on another thread.
 pub unsafe trait ByteSlice: Deref<Target = [u8]> + Sized {}
 
-#[allow(clippy::missing_safety_doc)] // TODO(fxbug.dev/99068)
+// Technically `ByteSliceMut` could be a safe trait since its safety invariants
+// are simply those of `ByteSlice`, but we make it `unsafe` to be conservative.
+#[allow(clippy::missing_safety_doc)]
```

Comment:
The `allow` here wigs me out a bit. We anticipate that outsiders might want to implement this trait, right? If so, we either need a safety comment or (better yet) we need to remove the `unsafe` modifier. Otherwise, we should seal the trait.

Reply:
I *believe* we can make this safe. Technically redundant, but maybe we could just have the safety invariant be that the type satisfies the stability property with respect to `DerefMut`? It's redundant in the sense that the `ByteSlice` safety docs already mention `DerefMut`.

Reply:
Discussed with @jswrenn offline: decided to just make `ByteSliceMut` a safe trait.